torch/csrc/utils/python_strings.h

```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include <Python.h>
+#include <stdexcept>
+#include "object_ptr.h"
+
+// Utilities for handling Python strings. Note that PyString, when defined, is
+// the same as PyBytes.
+
+// Returns true if obj is a bytes/str or unicode object
+// In existing CPython implementations, this does not require the GIL
+inline bool THPUtils_checkString(PyObject* obj) {
+  return PyBytes_Check(obj) || PyUnicode_Check(obj);
+}
+
+// Unpacks PyBytes (PyString) or PyUnicode as std::string
+// PyBytes are unpacked as-is. PyUnicode is unpacked as UTF-8.
+// NOTE: this method requires the GIL
+inline std::string THPUtils_unpackString(PyObject* obj) {
+  if (PyBytes_Check(obj)) {
+    size_t size = PyBytes_GET_SIZE(obj);
+    return std::string(PyBytes_AS_STRING(obj), size);
+  }
+  if (PyUnicode_Check(obj)) {
+#if PY_MAJOR_VERSION == 2
+    THPObjectPtr bytes = PyUnicode_AsUTF8String(obj);
+    if (!bytes) {
+      throw std::runtime_error("error unpacking string as utf-8");
+    }
+    size_t size = PyBytes_GET_SIZE(bytes.get());
+    return std::string(PyBytes_AS_STRING(bytes.get()), size);
+#else
+    Py_ssize_t size;
+    const char* data = PyUnicode_AsUTF8AndSize(obj, &size);
+    if (!data) {
+      throw std::runtime_error("error unpacking string as utf-8");
+    }
+    return std::string(data, (size_t)size);
+#endif
+  }
+  throw std::runtime_error("Could not unpack string");
```

Comment:
Maybe "unpackString expected bytes or unicode object"? This message is very uninformative.